what4/test/Bool.hs

```diff
@@ -0,0 +1,280 @@
+
+doGenExpr ::
+  ExprBuilder t st fs ->
+  GenT IO (BExpr (ExprBoundVar t BaseBoolType), Valuation t)
+doGenExpr sym =
+  let vars0 = Valuation Map.empty in
+  State.runStateT (genBExpr @(State.StateT _ (GenT IO)) (genVar @(GenT IO) sym)) vars0
+
+toSymExpr ::
+  IsExprBuilder sym =>
+  sym ->
+  -- | How to handle variables
+  (a -> IO (SymExpr sym BaseBoolType)) ->
+  BExpr a ->
+  IO (SymExpr sym BaseBoolType)
+toSymExpr sym var = go
+  where
+  go =
+    \case
+      Lit True -> pure (truePred sym)
+      Lit False -> pure (falsePred sym)
+      Var v -> var v
+      Not e -> notPred sym =<< go e
+      And l r -> do
+        l' <- go l
+        r' <- go r
+        andPred sym l' r'
+      Eq l r -> do
+        l' <- go l
+        r' <- go r
+        eqPred sym l' r'
+      Or l r -> do
+        l' <- go l
+        r' <- go r
+        orPred sym l' r'
+      Xor l r -> do
+        l' <- go l
+        r' <- go r
+        xorPred sym l' r'
+      Ite c l r -> do
+        c' <- go c
+        l' <- go l
+        r' <- go r
+        itePred sym c' l' r'
+
+-- | For use with 'toSymExpr', to leave variables uninterpreted
+uninterpVar :: ExprBoundVar t BaseBoolType -> Expr t BaseBoolType
+uninterpVar = BoundVarExpr
+
+eval :: Applicative f => (a -> f Bool) -> BExpr a -> f Bool
+eval var = go
+  where
+  ite c l r = if c then l else r
+  go =
+    \case
+      Lit True -> pure True
+      Lit False -> pure False
+      Var v -> var v
+      Not e -> not <$> go e
+      And l r -> (&&) <$> go l <*> go r
+      Eq l r -> (==) <$> go l <*> go r
+      Or l r -> (||) <$> go l <*> go r
+      Xor l r -> (/=) <$> go l <*> go r
+      Ite c l r -> ite <$> go c <*> go l <*> go r
+
+-- | For use with 'eval', to interpret variables
+getVar :: ExprBoundVar t BaseBoolType -> State.State (Valuation t) Bool
+getVar v = State.gets (getValue v)
+
+isNot :: Expr t BaseBoolType -> Bool
+isNot e =
+  case e of
+    AppExpr ae ->
+      case appExprApp ae of
+        NotPred {} -> True
+        _ -> False
+    _ -> False
+
+isNormalIte ::
+  ExprBuilder t st fs ->
+  Expr t BaseBoolType -> 
+  Expr t BaseBoolType -> 
+  Expr t BaseBoolType -> 
+  Either String ()
+isNormalIte sym c l r = do
+  isNormal sym c
+  isNormal sym l
+  isNormal sym r
+  when (isNot c) (Left "negated ite condition")
+  when (c == l) (Left "ite cond == LHS")
+  when (c == r) (Left "ite cond == RHS")
+  when (c == truePred sym) (Left "ite cond == true")
+  when (c == falsePred sym) (Left "ite cond == false")
+
+isNormalConjunct ::
+  ExprBuilder t st fs ->
+  Expr t BaseBoolType ->
+  BM.Polarity ->
+  Either String ()
+isNormalConjunct sym expr pol =
+  case expr of
+    BoolExpr {} -> Left "boolean literal inside conjunction"
+    BoundVarExpr {} -> Right ()
+    AppExpr ae ->
+      case appExprApp ae of
+        NotPred {} -> Left "not should be expressed via polarity"
+        -- This must be an OR, if it is an AND it should be combined with
+        -- its parent
```

Comment:
If we move the explanation for how boolean normalization works to the library Haddocks, then we could leave comments next to individual lines of code here in `test/Bool.hs` to justify why certain checks are performed. For instance, it's not obvious to me why this case should constitute an error, and I would have to trawl through the implementation in the library to figure out why this is a reasonable expectation. (Similarly for other checks in `test/Bool.hs`.)

Reply:
> to justify why certain checks are performed

Broadly, the justification is the same for all of them: to reduce the sizes of terms, to keep terms more concrete, and (in the service of the first two) to keep terms in normal forms (e.g., so that they can be compared for equality, and rewrite rules can assume normal forms as inputs). However, we could certainly note *how* each invariant contributes to these goals.

Reply:
Let's do this as part of #284.